[TaskValueSet(AndRun)] Use TASKname and/or VARname instead of TASKnr/VARnr

tonhuisman opened the pull request:
Implementation of allowing TASKname and/or VARname instead of TASKnr/VARnr in the TaskValueSet an TaskValueSetAndRun rules statements.
A substitution of the TASKname by TASKnr and VARname by VARnr is performed by looking up their values when the nr's are invalid.
Advantage of this method is, that it is a drop-in replacement of numbers by names, making rules 'backward compatible', but much more readable, and dummy variables no longer fixed-position in the devices list without changes in rules, for example (partial quote from the documentation):
```
 on sw1#state do
  if [dummy#var1]=0
    TaskValueSet dummy,var1,0
  else
    TaskValueSet dummy,var1,1
  endif
  gpio,16,[dummy#var1]
  gpio,13,[dummy#var1]
 endon
```
~~Not sure what issues will be resolved by this PR. (Couldn't find them easily)~~
Resolves: #1676

tonhuisman committed:
[TaskValueSet(AndRun)] Use TASKname and/or VARname instead of TASKnr/VARnr

tonhuisman committed:
[TaskValueSet(AndRun)] Minor documentation fixes

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

TD-er commented:
Not 100% sure this is the best syntax for referring task names and taskvars.

Right now we don't check for keywords in the names used.

`[taskname#varname]` is a reference to the value.
So I was thinking maybe we should have something similar for the reference to their task_index and var_index. This could then be replaced by `1,2` or something like that in the parser.
But maybe I'm a bit too worried here?

@uzi18 Do you have an opinion about this?

tonhuisman commented:
> Not 100% sure this is the best syntax for referring task names and taskvars.
> 
> Right now we don't check for keywords in the names used.

Well, this seemed to me the most 'in-place' upgrade possible, no need for pre-processing the script, no need for checking a variable number of arguments.
Why would using keywords in dummy/variable names be problematic, except for the predefined event names? I guess that should best be checked before saving the dummy variable. I could look into that...

uzi18 commented:
@TD-er Was thinking about it some time ago, but this syntax should be more explicit, maybe something like TaskValueSet@[dummy#var1],0
or simply:
TaskValueSet@dummy#var1,0
in fact it could be more complicated to code, but maybe we can teach command parser to check if @ exists after command name

TD-er commented:
Another option could be `[taskname@varname]`, which is then replaced in a pre-parser into something like this: `2,1`
The `[taskname#varname]` is also replaced by the pre-parser to replace all occurences of those into their sample values.
So all commands (not sure if there are many) that refer to a task number and var number are covered by this without any change in the code (maybe the check for number of arguments should be done after this replacement)

I am a bit worried about later commands that could be added that may also need similar code to support both names and numbers.
Also we can send commands via other means, like MQTT or UDP and I'm not sure if they should be using the names or the numbers.
Part of me is very much in favor of having the naming schema as it is particularly useful for communicating among nodes. But that also requires the other side to support this too.

But like I said, I'm not sure yet if we need to worry about this or not.

For sure it is a very good feature to have, but I would rather have had a good discussion about it before merge, instead of having to fix something later that may have been an obvious side effect we're not seeing right now.

uzi18 commented:
eventually:
let, taskname@varname, value
let, taskname#varname=value
or just:
taskname@varname=value
taskname#varname=value

TD-er commented:
Don't we have more commands acting on taskname/varname?
- TaskValueSet
- TaskValueSetAndRun
- TaskRun
- TaskEnable
- TaskDisable

Not sure if we have more.

Maybe we can also use it for commands of plugins that may have multiple instances.
So then we may need to have it as a prefix of the command?

uzi18 commented:
with my first idea with @ char as first we can  simply address both @task and @task#var

TaskValueSet@dummy#var1,0
TaskRun@dummy
TaskEnable@...
ControllerEnable@...

TD-er commented:
> eventually:
> let, taskname@varname, value
> let, taskname#varname=value
> or just:
> taskname@varname=value
> taskname#varname=value

I don't know if we also allow a single = for matching? (same as `==` )
If so, then I think it may be confusing to have the `operator=` like implementation for assignments.

tonhuisman committed:
[Task* commands] Add support for TASKname/VARname in addition to TASKnr/VARnr + documentation

tonhuisman commented:
Added the same support to the task commands I missed before (TaskClear, TaskRun, TaskToggle, TaskEnable, TaskDisable). And updated documentation.
So we can either merge or drop this for all task* commands after deciding what it should look like.

tonhuisman committed:
[Documentation] Add TaskValueSetAndRun

tonhuisman commented:
> 
> Another option could be `[taskname@varname]`, which is then replaced in a pre-parser into something like this: `2,1`

That sounds like the most feasible alternative to me, but...

> So all commands (not sure if there are many) that refer to a task number and var number are covered by this without any change in the code (maybe the check for number of arguments should be done after this replacement)

Thought about this too, and one of the reasons I chose the current approach is that now we have full control over what commands support tasknames/varnames, to avoid 'strange' behavior when [taskname@varname] is copied somewhere where [taskname#varname] was intended (or the other way around...).
Don't know if generic users would understand and appreciate the minimal difference.

TD-er commented:
There is also the risk of mistakes where # and @ are accidentally swapped.

So I do understand your PR very well and maybe it is one of the best solutions, but I just want to make sure we don't miss an obvious side effect.

uzi18 commented:
it is not a good idea it should be more generic, we have limited commands count to use with this syntax, so if we think about extra parameter we can use it as extra scenario

TD-er commented:
What if we take it from the other way around?
Right now we only have a limited number of commands addressing a task, but a lot of them could benefit from it. (the ones that allow multiple instances of the same plugin)
So then we only have to adapt the few commands that now do have a task parameter.
But that requires probably a command prefix instead of altering parameter matching.
It may also speed up command parsing as well, as  you no longer have to test against all tasks, but can directly address the task.

Without the addressing, it would default to the first enabled task that can handle the command.

So something like this kind of syntax:
```
@taskname:command,param1,....
@taskname#varname:command,param1,....
```
It also allows later for some extra options to address another node for example.
```
@[aa:bb:cc:dd:ee:ff]#taskname#varname:command,param1,....
@[unit:123]#taskname#varname:command,param1,....
@[192.168.1.2]#taskname#varname:command,param1,....
```

tonhuisman committed:
[Task* commands] Fix a few silly typos

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

clumsy-stefan commented:
If I understand this correct, it assumes that tasknames are unique, which currently is checked, but not enforced. also there are cases where duplicate tasknames make sense, which then using names instead of numbers would not work anymore…

just my 2cents… IMHO….

TD-er commented:
@clumsy-stefan The existing usage where you use numbers is still working.
Otherwise all existing rules out there will no longer work.

clumsy-stefan commented:
ok, and then qualify the names differntly than the numbers to ake sure its task nr 5 and not task name "5" I assume… sorry, probably I overread this somewhere…

tonhuisman commented:
> 
> ok, and then qualify the names differntly than the numbers to ake sure its task nr 5 and not task name "5" I assume… sorry, probably I overread this somewhere…

The check wether to use the name is based on the fact that a name doesn't properly convert to an integer. In that case the name-lookup is performed, so numbers get preference over the names (fully backward compatible), and no, devicenames like "5" won't be recognized, device number 5 will be addressed, not necessarily the device named "5" (unless of course you configured that as device 5... 😛)

clumsy-stefan commented:
Hmm.. so then with this change there is an implicit restriction on tasknames introduced, that task names should not be integers to avoid misinterpretation of rules… should just be documented then…

TD-er commented:
> Hmm.. so then with this change there is an implicit restriction on tasknames introduced, that task names should not be integers to avoid misinterpretation of rules… should just be documented then…

And/or warned for, the same way we warn for name duplicates

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

TD-er committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'feature/taskvalueset_by_name' of https://github.com/tonhuisman/ESPEasy-1 into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/taskvalueset_by_name

TD-er commented:
It looks like you messed up the PR again.

tonhuisman commented:
> It looks like you messed up the PR again.

That would be rather unexpected, looking at the list of changes, there's only 5 files changed, as would be expected 🤔

TD-er commented:
The `Misc.ino` file is back again.

tonhuisman committed:
[Merge] Remove (again)

tonhuisman commented:
Hm, I guess VSCode threw me off, I'm quite sure it had a state of Delete before I committed, but apparently not.
Fixed now.

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

TD-er commented on src/src/Commands/Tasks.cpp:
```diff
@@ -67,7 +67,7 @@ String Command_Task_ClearAll(struct EventStruct *event, const char *Line)
 }
 
-String Command_Task_EnableDisable(struct EventStruct *event, bool enable)
+String Command_Task_EnableDisable(struct EventStruct *event, bool enable, const char *Line)
```
Why use the `const char* Line` here?
It is not used, so that's why I did not add it as a function argument.

tonhuisman replied:
It is needed for this feature to work, transform a TaskName into a TaskIndex.

TD-er replied:
Ah OK.

Can't this be done before calling this function, right where the event is also filled with Var1...Var5 if they are numerical values?
That way you can have it in one place and set the `event->TaskIndex`

tonhuisman replied:
> Can't this be done before calling this function, right where the event is also filled with Var1...Var5

Then we would have to check first if the command is one of the Task* commands, as the `Command_Task_*` functions are called directly from the internal command processing. But this syntax isn't valid/applicable for other commands.

tonhuisman committed:
[Task* commands] Some code cleanup

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/taskvalueset_by_name

tonhuisman commented:
Because of recent changes I had to do a manual merge, and also got the opportunity to iron out some wrinkles 😉

tonhuisman committed:
[Task* commands] Code cleanup and some corrections

uzi18 commented:
@tonhuisman @TD-er and what about this proposition: https://github.com/letscontrolit/ESPEasy/pull/3098#issuecomment-637155906

tonhuisman commented:
My preference is to stay as close to the current script syntax as possible. And this is probably the smallest code to handle a simple command extension like this, for the referenced syntax I'd expect some sort of pre-parsing should be introduced, plus it's not very intuitive to me.

uzi18 commented:
In my opinion if we introduce something like this it should be extensible to address also external modules, to be consistent in the future.

uzi18 commented:
> quote from @TD-er 
> So something like this kind of syntax:
> 
> ```
> @taskname:command,param1,....
> @taskname#varname:command,param1,....
> ```
> 
> It also allows later for some extra options to address another node for example.
> 
> ```
> @[aa:bb:cc:dd:ee:ff]#taskname#varname:command,param1,....
> @[unit:123]#taskname#varname:command,param1,....
> @[192.168.1.2]#taskname#varname:command,param1,....
> ```

with such idea you can easily ommit address part and try send command to external unit
 
what if we introduce also this one, so we have: `@[unit:123]#taskname#varname:TaskValueSet dummy,1,1` so what to use dummy or taskname / varname or block such situation?

TD-er commented on src/src/Commands/Tasks.cpp:
```diff
@@ -39,12 +41,42 @@ bool validTaskVars(struct EventStruct *event, taskIndex_t& taskIndex, unsigned i
+    {
+      byte tmpVarNr = findDeviceValueIndexByName(valueName, event->Par1 - 1);
+      if (tmpVarNr != VARS_PER_TASK) {
+        event->Par2 = tmpVarNr + 1;
```
Shouldn't this also check the nr. of taskvalues for the specified task?

tonhuisman replied:
Yes, that would be more correct. Though this won't break anything. I'll have a look on adding that.

tonhuisman replied:
A few lines below there is a call to `validTaskVars` but that's also not checking for the number of values defined, and neither `taskValueSet` checks this... (but AFAIR there is a check somewhere in the code, I'll just have to find it 😃)

TD-er commented:
@uzi18 The suggested change made here by @tonhuisman looks like it is merely an optional replacement for a task name into a task index in the existing commands.
If we transition into prefixing the taskname and/or task variable name, we get 2 types for the same commands.

In another PR (which has been merged) we already allow to add a prefix for commands to specify a taskindex (by using a task name) for those commands that do not have a parameter for the task.
So essentially we already have somewhat of a mixed situation here.

Task specific commands like `taskrun`, `taskvalueset`, etc. are internal commands so they have become somewhat confusing whether to use `<taskname>.taskrun` or `taskrun,<taskname>` (or the task index of course)

Analog to the current `<taskname>.command` we can still later introduce a host/node prefix to address another node. (either via node nr, hostname, IP, MAC, etc.)

This makes me think we should perhaps not translate the taskname as a parameter for internal commands, but maybe convert those commands that accept a task index to allow to be addressed like this: `<taskname>.taskrun`

The "object oriented" notation of it does appeal to me as a programmer, but I do realize there are some people who may not think in an object oriented way.
So what does seem to be the least confusing notation?

tonhuisman commented:
> we can still later introduce a host/node prefix to address another node. (either via node nr, hostname, IP, MAC, etc.)

Isn't that what the `SendTo` command is intended for? I do appreciate that this currently only supports the ESPEasy p2p network over UDP (with its known limitations), but that's quite easy to expand into name/MAC resolution and/or TCP when using an ip-address (ipv4 or ipv6).

If we want this to be a prefix, then I'd suggest to use a syntax like `@<unit>@`, with or without an optional trailing period, where `<unit>` can be either a device-id, device-name, ip-address, MAC, etc. The prefix could then be stripped and the remaining command sent verbatim to the unit for further handling.

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman committed:
Merge branch 'mega' into feature/taskvalueset_by_name

tonhuisman commented:
I think this one can be merged, don't think a device addressing scheme as suggested/discussed, should be incorporated in this PR.
